File: pm_flutter/lib/models/blocker.dart
class Blocker {
  final int id;
  final int cardId;
  final String cardTitle;
  final String reason;
  final String priority;
  final String status;
  final BlockerUser reporter;
  final BlockerUser? assignee;
  final String? resolutionNote;
  final int timeBlockedHours;
  final bool isOverdue;
  final DateTime createdAt;
  final DateTime? resolvedAt;
  final List<BlockerComment>? comments;

  Blocker({
    required this.id,
    required this.cardId,
    required this.cardTitle,
    required this.reason,
    required this.priority,
    required this.status,
    required this.reporter,
    this.assignee,
    this.resolutionNote,
    required this.timeBlockedHours,
    required this.isOverdue,
    required this.createdAt,
    this.resolvedAt,
    this.comments,
  });

  factory Blocker.fromJson(Map<String, dynamic> json) {
    return Blocker(
      id: json['id'] as int,
      cardId: json['card_id'] as int,
      cardTitle: json['card_title'] as String,
      reason: json['reason'] as String,
      priority: json['priority'] as String,
      status: json['status'] as String,
      reporter: BlockerUser.fromJson(json['reporter'] as Map<String, dynamic>),
      assignee: json['assignee'] != null
          ? BlockerUser.fromJson(json['assignee'] as Map<String, dynamic>)
          : null,
      resolutionNote: json['resolution_note'] as String?,
      timeBlockedHours: json['time_blocked_hours'] as int,
      isOverdue: json['is_overdue'] as bool,
      createdAt: DateTime.parse(json['created_at'] as String),
      resolvedAt: json['resolved_at'] != null
          ? DateTime.parse(json['resolved_at'] as String)
          : null,
      comments: json['comments'] != null
          ? (json['comments'] as List)
              .map((c) => BlockerComment.fromJson(c as Map<String, dynamic>))
              .toList()
          : null,
    );
  }

  String get priorityLabel {
    switch (priority) {
      case 'critical':
        return 'Critical';
      case 'high':
        return 'High';
      case 'medium':
        return 'Medium';
      case 'low':
        return 'Low';
      default:
        return priority;
    }
  }

  String get statusLabel {
    switch (status) {
      case 'reported':
        return 'Reported';
      case 'assigned':
        return 'Assigned';
      case 'in_progress':
        return 'In Progress';
      case 'resolved':
        return 'Resolved';
      default:
        return status;
    }
  }
}

class BlockerUser {
  final int id;
  final String name;
  final String? avatar;

  BlockerUser({
    required this.id,
    required this.name,
    this.avatar,
  });

  factory BlockerUser.fromJson(Map<String, dynamic> json) {
    return BlockerUser(
      id: json['id'] as int,
      name: json['name'] as String,
      avatar: json['avatar'] as String?,
    );
  }
}

class BlockerComment {
  final int id;
  final BlockerUser user;
  final String comment;
  final DateTime createdAt;

  BlockerComment({
    required this.id,
    required this.user,
    required this.comment,
    required this.createdAt,
  });

  factory BlockerComment.fromJson(Map<String, dynamic> json) {
    return BlockerComment(
      id: json['id'] as int,
      user: BlockerUser.fromJson(json['user'] as Map<String, dynamic>),
      comment: json['comment'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

File: pm/app/Models/Blocker.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Blocker extends Model
{
    use HasFactory;

    protected $fillable = [
        'card_id',
        'reporter_id',
        'reason',
        'priority',
        'status',
        'assigned_to',
        'resolution_note',
        'resolved_at',
    ];

    protected $casts = [
        'resolved_at' => 'datetime',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    /**
     * Get the task/card that is blocked
     */
    public function card(): BelongsTo
    {
        return $this->belongsTo(Card::class, 'card_id', 'card_id');
    }

    /**
     * Get the user who reported the blocker
     */
    public function reporter(): BelongsTo
    {
        return $this->belongsTo(User::class, 'reporter_id', 'user_id');
    }

    /**
     * Get the user assigned to help
     */
    public function assignee(): BelongsTo
    {
        return $this->belongsTo(User::class, 'assigned_to', 'user_id');
    }

    /**
     * Get comments for this blocker
     */
    public function comments(): HasMany
    {
        return $this->hasMany(BlockerComment::class)->orderBy('created_at', 'asc');
    }

    /**
     * Scope for active blockers
     */
    public function scopeActive($query)
    {
        return $query->whereIn('status', ['reported', 'assigned', 'in_progress']);
    }

    /**
     * Scope for resolved blockers
     */
    public function scopeResolved($query)
    {
        return $query->where('status', 'resolved');
    }

    /**
     * Scope for high priority
     */
    public function scopeHighPriority($query)
    {
        return $query->whereIn('priority', ['high', 'critical']);
    }

    /**
     * Get time blocked in hours
     */
    public function getTimeBlockedAttribute()
    {
        $end = $this->resolved_at ?? now();
        return $this->created_at->diffInHours($end);
    }

    /**
     * Check if blocker is overdue (>24 hours without resolution)
     */
    public function getIsOverdueAttribute()
    {
        if ($this->status === 'resolved') {
            return false;
        }
        return $this->created_at->addHours(24)->isPast();
    }
}
